include/seastar/core/sstring.hh

```diff
@@ -656,6 +657,39 @@ operator>>(std::basic_istream<char_type, char_traits>& is,
     return is;
 }
 
+/// an implementation of std::back_insert_iterator<basic_sstring<...>>
+///
+/// So we can use fmt::format_to() to format given args to a preallocated sstring with
+/// enough space. Please use `fmt::formatted_size()` to calculate the size of the
+/// preallocated sstring.
+template <typename char_type, typename size_type, size_type max_size, bool NulTerminate>
+class sstring_back_inserter {
+    char* begin;
+    const char* end;
+    using container_type = basic_sstring<char_type, size_type, max_size, NulTerminate>;
+public:
+    using iterator_category = std::output_iterator_tag;
+    using value_type = void;
+    using difference_type = ptrdiff_t;
+    using pointer = void;
+    using reference = void;
+    explicit constexpr sstring_back_inserter(container_type& c)
+    : begin(c.begin()), end(c.end()) {}
+    constexpr sstring_back_inserter operator=(char c) {
+        assert(begin != end);
+        *begin++ = c;
```

Comment:
How does this work? The string might need to be reallocated if you overflow its capacity.

sstring doesn't even store capacity, so it will need to be reallocated each character.

Reply:
yes, this relies on a `sstring` whose size is estimated using something like `fmt::formatted_size()` .

Reply:
Well, we can't expose a generic back_inserter if it works for just one caller.

Reply:
i don't follow you here. what gets exposed is not a generic `std::back_inserter`. it is a partial specialization of `std::back_inserter`. and what do you mean by "just one caller".